<a href="https://colab.research.google.com/github/norayehia/ETL-Employee/blob/main/Employeedata.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
def get_employees_df():

  return pd.read_csv(
      "https://gist.githubusercontent.com/kevin336/acbb2271e66c10a5b73aacf82"
        "ca82784/raw/e38afe62e088394d61ed30884dd50a6826eee0a8/employees.csv"
  )
def get_departments_df():
  dep_df = pd.read_csv(
      "https://gist.githubusercontent.com/kevin336/5ea0e96813aa88871c20d315b5"
        "bf445c/raw/d8fcf5c2630ba12dd8802a2cdd5480621b6a0ea6/departments.csv"
  )
  dep_df = dep_df.rename(columns={"DEPARTMENT_ID": "DEPARTMENT_IDENTIFIER"})

  return dep_df

employees = get_employees_df()
departments = get_departments_df()


In [4]:
departments.head(4)

DEPARTMENT_IDENTIFIER  DEPARTMENT_NAME MANAGER_ID  LOCATION_ID
0                     10   Administration        200         1700
1                     20        Marketing        201         1800
2                     30       Purchasing        114         1700
3                     40  Human Resources        203         2400

In [2]:
employees.head(3)

EMPLOYEE_ID FIRST_NAME LAST_NAME     EMAIL  PHONE_NUMBER  HIRE_DATE  \
0          198     Donald  OConnell  DOCONNEL  650.507.9833  21-JUN-07   
1          199    Douglas     Grant    DGRANT  650.507.9844  13-JAN-08   
2          200   Jennifer    Whalen   JWHALEN  515.123.4444  17-SEP-03   

     JOB_ID  SALARY COMMISSION_PCT MANAGER_ID  DEPARTMENT_ID  
0  SH_CLERK    2600             -         124             50  
1  SH_CLERK    2600             -         124             50  
2   AD_ASST    4400             -         101             10

# 1. Please calculate the average, median, lower and upper quartiles of an employees' salaries.

# 2. Please calculate the average salary per department. Please include the department name in the results.

# 3. Please create a new column named `SALARY_CATEGORY` with value "low" when the salary is lower than average and "high" if is it higher or equal.

# 4. Please create another column named `SALARY_CATEGORY_AMONG_DEPARTMENT` with value "low" when the employee salary is lower than average in his / her department and "high" in the other case.


# 5. Please filter the dataframe `employees` to include only the rows where `DEPARTMENT_ID` equals to 20. Assign the result to new variable.

# 6. Please increase the salary by 10% for all employees working at the department 20.

# 7. Please check if any of the `PHONE_NUMBER` column values are empty.


#1

In [10]:
#average
mSalary = employees['SALARY'].mean()
#median
mdSalary = employees['SALARY'].median()
print(mSalary)
print(mdSalary)

6182.32
4600.0


In [11]:
#lower .25and upper quartiles .75 below of an employees' salaries.
lowerquartile = employees['SALARY'].quantile(0.25)
upperquartile = employees['SALARY'].quantile(0.75)

In [12]:
print(lowerquartile)
print(upperquartile)

2725.0
8150.0


#2

In [14]:
# Merge employees with departments on the department identifier
mergeddf = pd.merge(employees, departments, left_on='DEPARTMENT_ID', right_on='DEPARTMENT_IDENTIFIER')


In [15]:
mergeddf.head(4)

EMPLOYEE_ID FIRST_NAME LAST_NAME     EMAIL  PHONE_NUMBER  HIRE_DATE  \
0          198     Donald  OConnell  DOCONNEL  650.507.9833  21-JUN-07   
1          199    Douglas     Grant    DGRANT  650.507.9844  13-JAN-08   
2          120    Matthew     Weiss    MWEISS  650.123.1234  18-JUL-04   
3          121       Adam     Fripp    AFRIPP  650.123.2234  10-APR-05   

     JOB_ID  SALARY COMMISSION_PCT MANAGER_ID_x  DEPARTMENT_ID  \
0  SH_CLERK    2600             -           124             50   
1  SH_CLERK    2600             -           124             50   
2    ST_MAN    8000             -           100             50   
3    ST_MAN    8200             -           100             50   

   DEPARTMENT_IDENTIFIER DEPARTMENT_NAME MANAGER_ID_y  LOCATION_ID  
0                     50        Shipping          121         1500  
1                     50        Shipping          121         1500  
2                     50        Shipping          121         1500  
3                     50        Shipping          121         1500

In [17]:
# Group by department name and calculate the average salary
average_salary = mergeddf.groupby('DEPARTMENT_NAME' )['SALARY'].mean().reset_index()
average_salary

DEPARTMENT_NAME        SALARY
0        Accounting  10154.000000
1    Administration   4400.000000
2         Executive  19333.333333
3           Finance   8601.333333
4   Human Resources   6500.000000
5                IT   5760.000000
6         Marketing   9500.000000
7  Public Relations  10000.000000
8        Purchasing   4150.000000
9          Shipping   3721.739130

#3

In [18]:


# Create a column SALARY_CATEGORY ,msalary above is average salary
employees['SALARY_CATEGORY'] = employees['SALARY'].apply(lambda x: 'low' if x < mSalary else 'high')


In [19]:
employees.head(3)

EMPLOYEE_ID FIRST_NAME LAST_NAME     EMAIL  PHONE_NUMBER  HIRE_DATE  \
0          198     Donald  OConnell  DOCONNEL  650.507.9833  21-JUN-07   
1          199    Douglas     Grant    DGRANT  650.507.9844  13-JAN-08   
2          200   Jennifer    Whalen   JWHALEN  515.123.4444  17-SEP-03   

     JOB_ID  SALARY COMMISSION_PCT MANAGER_ID  DEPARTMENT_ID SALARY_CATEGORY  
0  SH_CLERK    2600             -         124             50             low  
1  SH_CLERK    2600             -         124             50             low  
2   AD_ASST    4400             -         101             10             low

#4

Please create another column named SALARY_CATEGORY_AMONG_DEPARTMENT with value "low" when the employee salary is lower than average in his / her department and "high" in the other case.

In [21]:
#Group by department name and calculate the average salary
#average_salary = merged_df.groupby('DEPARTMENT_NAME')['SALARY'].mean().reset_index()

# Rename the column in average_salary to avoid conflicts
average_salary = average_salary.rename(columns={'SALARY': 'AVGDEPARTMENTSALARY'})

average_salary

DEPARTMENT_NAME  AVGDEPARTMENTSALARY
0        Accounting         10154.000000
1    Administration          4400.000000
2         Executive         19333.333333
3           Finance          8601.333333
4   Human Resources          6500.000000
5                IT          5760.000000
6         Marketing          9500.000000
7  Public Relations         10000.000000
8        Purchasing          4150.000000
9          Shipping          3721.739130

In [23]:
# Merge the average salary back into the main DataFrame to merged_df
merged_df = pd.merge(mergeddf, average_salary, on='DEPARTMENT_NAME', how='left')

merged_df

EMPLOYEE_ID   FIRST_NAME    LAST_NAME     EMAIL  PHONE_NUMBER  HIRE_DATE  \
0           198       Donald     OConnell  DOCONNEL  650.507.9833  21-JUN-07   
1           199      Douglas        Grant    DGRANT  650.507.9844  13-JAN-08   
2           120      Matthew        Weiss    MWEISS  650.123.1234  18-JUL-04   
3           121         Adam        Fripp    AFRIPP  650.123.2234  10-APR-05   
4           122        Payam     Kaufling  PKAUFLIN  650.123.3234  01-MAY-03   
5           123       Shanta      Vollman  SVOLLMAN  650.123.4234  10-OCT-05   
6           124        Kevin      Mourgos  KMOURGOS  650.123.5234  16-NOV-07   
7           125        Julia        Nayer    JNAYER  650.124.1214  16-JUL-05   
8           126        Irene  Mikkilineni  IMIKKILI  650.124.1224  28-SEP-06   
9           127        James       Landry   JLANDRY  650.124.1334  14-JAN-07   
10          128       Steven       Markle   SMARKLE  650.124.1434  08-MAR-08   
11          129        Laura       Bissot   LBISSOT  650.124.5234  20-AUG-05   
12          130        Mozhe     Atkinson  MATKINSO  650.124.6234  30-OCT-05   
13          131        James       Marlow   JAMRLOW  650.124.7234  16-FEB-05   
14          132           TJ        Olson   TJOLSON  650.124.8234  10-APR-07   
15          133        Jason       Mallin   JMALLIN  650.127.1934  14-JUN-04   
16          134      Michael       Rogers   MROGERS  650.127.1834  26-AUG-06   
17          135           Ki          Gee      KGEE  650.127.1734  12-DEC-07   
18          136        Hazel   Philtanker  HPHILTAN  650.127.1634  06-FEB-08   
19          137       Renske       Ladwig   RLADWIG  650.121.1234  14-JUL-03   
20          138      Stephen       Stiles   SSTILES  650.121.2034  26-OCT-05   
21          139         John          Seo      JSEO  650.121.2019  12-FEB-06   
22          140       Joshua        Patel    JPATEL  650.121.1834  06-APR-06   
23          200     Jennifer       Whalen   JWHALEN  515.123.4444  17-SEP-03   
24          201      Michael    Hartstein  MHARTSTE  515.123.5555  17-FEB-04   
25          202          Pat          Fay      PFAY  603.123.6666  17-AUG-05   
26          203        Susan       Mavris   SMAVRIS  515.123.7777  07-JUN-02   
27          204      Hermann         Baer     HBAER  515.123.8888  07-JUN-02   
28          205      Shelley      Higgins  SHIGGINS  515.123.8080  07-JUN-02   
29          206      William        Gietz    WGIETZ  515.123.8181  07-JUN-02   
30          100       Steven         King     SKING  515.123.4567  17-JUN-03   
31          101        Neena      Kochhar  NKOCHHAR  515.123.4568  21-SEP-05   
32          102          Lex      De Haan   LDEHAAN  515.123.4569  13-JAN-01   
33          103    Alexander       Hunold   AHUNOLD  590.423.4567  03-JAN-06   
34          104        Bruce        Ernst    BERNST  590.423.4568  21-MAY-07   
35          105        David       Austin   DAUSTIN  590.423.4569  25-JUN-05   
36          106        Valli    Pataballa  VPATABAL  590.423.4560  05-FEB-06   
37          107        Diana      Lorentz  DLORENTZ  590.423.5567  07-FEB-07   
38          108        Nancy    Greenberg  NGREENBE  515.124.4569  17-AUG-02   
39          109       Daniel       Faviet   DFAVIET  515.124.4169  16-AUG-02   
40          110         John         Chen     JCHEN  515.124.4269  28-SEP-05   
41          111       Ismael      Sciarra  ISCIARRA  515.124.4369  30-SEP-05   
42          112  Jose Manuel        Urman   JMURMAN  515.124.4469  07-MAR-06   
43          113         Luis         Popp     LPOPP  515.124.4567  07-DEC-07   
44          114          Den     Raphaely  DRAPHEAL  515.127.4561  07-DEC-02   
45          115    Alexander         Khoo     AKHOO  515.127.4562  18-MAY-03   
46          116       Shelli        Baida    SBAIDA  515.127.4563  24-DEC-05   
47          117        Sigal       Tobias   STOBIAS  515.127.4564  24-JUL-05   
48          118          Guy       Himuro   GHIMURO  515.127.4565  15-NOV-06   
49  

In [25]:
# Create the SALARY_CATEGORY_AMONG_DEPARTMENT column
employees['SALARY_CATEGORY_AMONG_DEPARTMENT'] = merged_df.apply(
    lambda row: 'low' if row['SALARY'] < row['AVGDEPARTMENTSALARY'] else 'high', axis=1
)



In [26]:
employees.head(3)

EMPLOYEE_ID FIRST_NAME LAST_NAME     EMAIL  PHONE_NUMBER  HIRE_DATE  \
0          198     Donald  OConnell  DOCONNEL  650.507.9833  21-JUN-07   
1          199    Douglas     Grant    DGRANT  650.507.9844  13-JAN-08   
2          200   Jennifer    Whalen   JWHALEN  515.123.4444  17-SEP-03   

     JOB_ID  SALARY COMMISSION_PCT MANAGER_ID  DEPARTMENT_ID SALARY_CATEGORY  \
0  SH_CLERK    2600             -         124             50             low   
1  SH_CLERK    2600             -         124             50             low   
2   AD_ASST    4400             -         101             10             low   

  SALARY_CATEGORY_AMONG_DEPARTMENT  
0                              low  
1                              low  
2                             high

#5

5. Please filter the dataframe employees to include only the rows where DEPARTMENT_ID equals to 20. Assign the result to new variable.

In [31]:
# Filter the DataFrame to include only rows where DEPARTMENT_ID equals 20
employeesfilter = employees[employees['DEPARTMENT_ID'] == 20]

# Display the filtered DataFrame
employeesfilter


EMPLOYEE_ID FIRST_NAME  LAST_NAME     EMAIL  PHONE_NUMBER  HIRE_DATE  \
3          201    Michael  Hartstein  MHARTSTE  515.123.5555  17-FEB-04   
4          202        Pat        Fay      PFAY  603.123.6666  17-AUG-05   

   JOB_ID  SALARY COMMISSION_PCT MANAGER_ID  DEPARTMENT_ID SALARY_CATEGORY  \
3  MK_MAN   13000             -         100             20            high   
4  MK_REP    6000             -         201             20             low   

  SALARY_CATEGORY_AMONG_DEPARTMENT  
3                             high  
4                             high

#6

6. Please increase the salary by 10% for all employees working at the department 20.

In [33]:
employeesfilter['SALARY']=employeesfilter['SALARY']*1.1

<ipython-input-33-a02bb851619f>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  employeesfilter['SALARY']=employeesfilter['SALARY']*1.1


In [34]:
employeesfilter['SALARY']

3    14300.0
4     6600.0
Name: SALARY, dtype: float64

In [35]:


# Safely increase the salary by 10% for  employeestable
employees.loc[employees['DEPARTMENT_ID'] == 20, 'SALARY'] *= 1.10



<ipython-input-35-3dc3593bd7b0>:2: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value '[14300.  6600.]' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.
  employees.loc[employees['DEPARTMENT_ID'] == 20, 'SALARY'] *= 1.10


In [36]:
employees.head(4)

EMPLOYEE_ID FIRST_NAME  LAST_NAME     EMAIL  PHONE_NUMBER  HIRE_DATE  \
0          198     Donald   OConnell  DOCONNEL  650.507.9833  21-JUN-07   
1          199    Douglas      Grant    DGRANT  650.507.9844  13-JAN-08   
2          200   Jennifer     Whalen   JWHALEN  515.123.4444  17-SEP-03   
3          201    Michael  Hartstein  MHARTSTE  515.123.5555  17-FEB-04   

     JOB_ID   SALARY COMMISSION_PCT MANAGER_ID  DEPARTMENT_ID SALARY_CATEGORY  \
0  SH_CLERK   2600.0             -         124             50             low   
1  SH_CLERK   2600.0             -         124             50             low   
2   AD_ASST   4400.0             -         101             10             low   
3    MK_MAN  14300.0             -         100             20            high   

  SALARY_CATEGORY_AMONG_DEPARTMENT  
0                              low  
1                              low  
2                             high  
3                             high

#You can see Employee with Id 201 its salary be 14300

#7.

7. Please check if any of the PHONE_NUMBER column values are empty.

In [37]:
employees['PHONE_NUMBER'].isnull().sum()

0

0 this mean no empty values